assets/components/dialog/dialog.jsx

```diff
@@ -0,0 +1,115 @@
+// @flow
+
+// ----- Imports ----- //
+
+import React, { Component, type Node } from 'react';
+
+import { type Option } from 'helpers/types/option';
+import { classNameWithModifiers } from 'helpers/utilities';
+
+import './dialog.scss';
+
+
+// ----- Props ----- //
+
+export type PropTypes = {|
+  onStatusChange: (boolean) => void,
+  modal: boolean,
+  open: boolean,
+  'aria-label': Option<string>,
+  dismissOnBackgroundClick: boolean,
+  children: Node
+|};
+
+
+// ----- Component ----- //
+
+class Dialog extends Component<PropTypes> {
+
+  static defaultProps = {
+    onStatusChange: () => {},
+    modal: true,
+    open: false,
+    dismissOnBackgroundClick: false,
+  }
+
+  componentDidMount() {
+    if (this.props.open) {
+      this.open();
+    }
+  }
+
+  componentDidUpdate(prevProps: PropTypes) {
+    if (prevProps.open === true && this.props.open === false) {
+      this.close();
+    } else if (prevProps.open === false && this.props.open === true) {
+      this.open();
+    }
+  }
+
+  open() {
+    if (this.ref && this.ref.showModal) {
+      if (this.props.modal) {
+        this.ref.showModal();
+      } else {
+        this.ref.show();
+      }
+    }
+    requestAnimationFrame(() => {
+      if (this.ref) {
+        this.ref.focus();
+      }
+    });
+  }
+
+  close() {
+    if (this.ref && this.ref.close) {
+      this.ref.close();
+    }
+  }
+
+  ref: ?(HTMLDialogElement & {focus: Function});
+
+  render() {
+    const {
+      open, modal, children, onStatusChange, dismissOnBackgroundClick, ...otherProps
+    } = this.props;
+
+    return (
+      <dialog // eslint-disable-line jsx-a11y/no-redundant-roles
+        className={classNameWithModifiers('component-dialog', [modal ? 'modal' : null, open ? 'open' : null])}
+        aria-modal={modal}
+        aria-hidden={!open}
+        tabIndex="-1"
```

Comment:
So, since browser support for `dialog` is fiddly at best I have decided to implement focus management cheaply by just focusing the whole dialog after it opens – this seems to net sensible results in all tested browsers.

The native behaviour would be to focus on the first interactive element – however that requires more DOM fiddling, i'm not against exploring that solution but I'm not confident it's as portable as focusing the whole thing :(

I found healthy debate [here](https://github.com/whatwg/html/issues/1929) about this all, my takeaways are:
- The guidance refers more about a numberical tabindex than the magical `-1` and `0` (because `dialog`s live outside of the element to a level)
- Some user agents make `dialog` focusable? (thinking about using `tabindex=0`
- Focusing the dialog itself vs focusing the first interactive element is an open discussion?

Reply:
Awesome, thanks for sharing the discussion, really interesting! 

The guidance we received from the accessibility consultant:

- if there's an obvious candidate to receive focus in the dialog, use that, OTHERWISE
- provide focusable empty divs at the top and bottom of the dialog and use JavaScript to return focus to the top div when tabbing out of the bottom div (and vice versa). 

I can link you the notes he provided if you need it. But there's more than one way to peel a potato, and as long as the approach is tested, it's all good! 👍

Reply:
we do have a naively implemented oneline focus trap at the end but not one at the top 😢 I'm not sure how to implement both sides without a lot of code. it's something i want to look into though!